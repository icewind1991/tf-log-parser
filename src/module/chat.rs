use crate::common::{SubjectData, SubjectId};
use crate::module::EventHandler;
use crate::raw_event::{RawEvent, RawEventType};
use crate::SubjectMap;
use std::convert::Infallible;
use steamid_ng::SteamID;

struct BareChatMessage {
    pub time: u32,
    pub subject: SubjectId,
    pub message: String,
    pub chat_type: ChatType,
}

pub struct ChatMessage {
    pub time: u32,
    pub name: String,
    pub steam_id: SteamID,
    pub message: String,
    pub chat_type: ChatType,
}

impl ChatMessage {
    fn from_bare(bare: BareChatMessage, subjects: &SubjectMap) -> Self {
        let (name, steam_id) = match &subjects[bare.subject] {
            SubjectData::Player { name, steam_id, .. } => (name.clone(), steam_id.clone()),
            _ => {
                unreachable!("only player messages are added");
            }
        };
        ChatMessage {
            time: bare.time,
            name,
            steam_id,
            message: bare.message,
            chat_type: bare.chat_type,
        }
    }
}

pub enum ChatType {
    All,
    Team,
}

#[derive(Default)]
pub struct ChatHandler(Vec<BareChatMessage>);

impl EventHandler for ChatHandler {
    type Output = Vec<ChatMessage>;
    type Error = Infallible;

    fn does_handle(&self, ty: RawEventType) -> bool {
        matches!(ty, RawEventType::SayTeam | RawEventType::Say)
    }

    fn handle(
        &mut self,
        time: u32,
        subject: SubjectId,
        event: &RawEvent,
    ) -> Result<(), Infallible> {
        if !matches!(subject, SubjectId::Player(_)) {
            return Ok(());
        }
        match event.ty {
            RawEventType::SayTeam => self.0.push(BareChatMessage {
                time,
                subject,
                message: event.params.trim_matches('"').into(),
                chat_type: ChatType::Team,
            }),
            RawEventType::Say => self.0.push(BareChatMessage {
                time,
                subject,
                message: event.params.trim_matches('"').into(),
                chat_type: ChatType::All,
            }),
            _ => {}
        }
        Ok(())
    }

    fn finish(self, subjects: &SubjectMap) -> Self::Output {
        self.0
            .into_iter()
            .map(|bare| ChatMessage::from_bare(bare, subjects))
            .collect()
    }
}
