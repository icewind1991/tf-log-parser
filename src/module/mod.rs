use crate::common::SubjectId;
use crate::event::GameEvent;
use crate::raw_event::RawEventType;
use crate::SubjectMap;
pub use chat::{ChatHandler, ChatMessage, ChatType};
pub use healspread::HealSpreadHandler;
pub use lobbysettings::{
    LobbySettingsError, LobbySettingsHandler, Location, Settings as LobbySettings,
};
pub use medicstats::{MedicStats, MedicStatsHandler};
use std::convert::Infallible;
use std::error::Error;
use std::fmt::Debug;
use thiserror::Error;

mod chat;
mod healspread;
mod lobbysettings;
mod medicstats;

pub trait EventHandler: Default {
    type Output;
    type Error: Error + Debug;

    fn does_handle(&self, ty: RawEventType) -> bool;

    fn handle(
        &mut self,
        time: u32,
        subject: SubjectId,
        event: &GameEvent,
    ) -> Result<(), Self::Error>;

    fn finish(self, subjects: &SubjectMap) -> Self::Output;
}

#[derive(Default)]
pub struct HandlerStack<Head, Tail> {
    head: Head,
    tail: Tail,
}

impl<Head: EventHandler, Tail: EventHandler> EventHandler for HandlerStack<Head, Tail> {
    type Output = (Head::Output, Tail::Output);
    type Error = EitherError<Head::Error, Tail::Error>;

    fn does_handle(&self, ty: RawEventType) -> bool {
        self.head.does_handle(ty) || self.tail.does_handle(ty)
    }

    fn handle(
        &mut self,
        time: u32,
        subject: SubjectId,
        event: &GameEvent,
    ) -> Result<(), Self::Error> {
        self.head
            .handle(time, subject, event)
            .map_err(|e| EitherError::E1(e))?;
        self.tail
            .handle(time, subject, event)
            .map_err(|e| EitherError::E2(e))?;
        Ok(())
    }

    fn finish(self, subjects: &SubjectMap) -> Self::Output {
        (self.head.finish(subjects), self.tail.finish(subjects))
    }
}

#[derive(Debug, Error)]
pub enum EitherError<E1: Error, E2: Error> {
    #[error("{0}")]
    E1(E1),
    #[error("{0}")]
    E2(E2),
}

/// A handler that doesn't stop the parsing on failure
pub enum OptionalHandler<Handler: EventHandler> {
    Active(Handler),
    Failed(Handler::Error),
}

impl<Handler: EventHandler> Default for OptionalHandler<Handler> {
    fn default() -> Self {
        OptionalHandler::Active(Handler::default())
    }
}

impl<Handler: EventHandler> EventHandler for OptionalHandler<Handler> {
    type Output = Result<Handler::Output, Handler::Error>;
    type Error = Infallible;

    fn does_handle(&self, ty: RawEventType) -> bool {
        match self {
            OptionalHandler::Active(handler) => handler.does_handle(ty),
            OptionalHandler::Failed(_) => false,
        }
    }

    fn handle(
        &mut self,
        time: u32,
        subject: SubjectId,
        event: &GameEvent,
    ) -> Result<(), Self::Error> {
        let res = if let OptionalHandler::Active(handler) = self {
            handler.handle(time, subject, event)
        } else {
            Ok(())
        };

        if let Err(e) = res {
            dbg!(&e);
            *self = OptionalHandler::Failed(e)
        }
        Ok(())
    }

    fn finish(self, subjects: &SubjectMap) -> Self::Output {
        match self {
            OptionalHandler::Active(handler) => Ok(handler.finish(subjects)),
            OptionalHandler::Failed(e) => Err(e),
        }
    }
}
