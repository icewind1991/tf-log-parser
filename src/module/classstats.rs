use crate::common::{Class, ClassMap, SubjectId};
use crate::event::{DamageEvent, GameEvent, RoleChangeEvent, SpawnEvent};
use crate::module::EventHandler;
use crate::raw_event::{RawEventType, RawSubject};
use crate::{EventMeta, SubjectData, SubjectMap};
use serde::Serialize;
use std::collections::BTreeMap;

#[derive(Debug, Serialize, Default, PartialEq)]
pub struct ClassStats {
    pub kills: ClassMap<u8>,
    pub deaths: ClassMap<u8>,
    pub assists: ClassMap<u8>,
    pub damage: ClassMap<u16>,
}

#[derive(Default)]
pub struct ClassStatsHandler {
    active: bool,
    data: BTreeMap<SubjectId, ClassStatData>,
}

#[derive(Default)]
pub struct ClassStatData {
    class: Class,
    deaths: ClassMap<u8>,
}

impl ClassStatsHandler {
    fn get_class(&self, subject: &RawSubject) -> Option<Class> {
        subject
            .id()
            .ok()
            .and_then(|id| self.data.get(&id))
            .map(|data| data.class)
    }

    fn data_mut(&mut self, id: SubjectId) -> &mut ClassStatData {
        self.data.entry(id).or_default()
    }
}

impl EventHandler for ClassStatsHandler {
    type GlobalOutput = ();
    type PerSubjectData = ClassStats;
    type PerSubjectOutput = ClassStats;

    fn does_handle(ty: RawEventType) -> bool {
        matches!(
            ty,
            RawEventType::Killed
                | RawEventType::KillAssist
                | RawEventType::Damage
                | RawEventType::Spawned
                | RawEventType::RoleChange
                | RawEventType::RoundWin
                | RawEventType::RoundStart
        )
    }

    fn handle(
        &mut self,
        _meta: &EventMeta,
        subject: SubjectId,
        subject_data: &mut Self::PerSubjectData,
        event: &GameEvent,
    ) {
        match event {
            GameEvent::Spawned(SpawnEvent { class: Some(class) })
            | GameEvent::RoleChange(RoleChangeEvent { class: Some(class) }) => {
                self.data_mut(subject).class = *class;
            }
            GameEvent::RoundStart => {
                self.active = true;
            }
            GameEvent::RoundWin(_) => {
                self.active = false;
            }
            GameEvent::Killed(kill) if self.active => {
                if let Ok(target) = kill.target.id() {
                    let subject_class = self.data.get(&subject).map(|data| data.class);
                    let target_data = self.data_mut(target);
                    if let Some(subject_class) = subject_class {
                        target_data.deaths[subject_class] += 1;
                    }
                    subject_data.kills[target_data.class] += 1;
                }
            }
            GameEvent::KillAssist(assist) if self.active => {
                if let Some(target_class) = self.get_class(&assist.target) {
                    subject_data.assists[target_class] += 1;
                }
            }
            GameEvent::Damage(DamageEvent {
                damage: Some(damage),
                target,
                ..
            }) if self.active && damage > &0 && damage < &1500 => {
                if let Some(target_class) = self.get_class(target) {
                    subject_data.damage[target_class] =
                        subject_data.damage[target_class].saturating_add(*damage as u16);
                }
            }
            _ => {}
        }
    }

    fn finish_global(self, _subjects: &SubjectMap) -> Self::GlobalOutput {}

    fn finish_per_subject(
        &mut self,
        subject: &SubjectData,
        mut data: Self::PerSubjectData,
    ) -> Self::PerSubjectOutput {
        data.deaths = self.data.remove(&subject.id()).unwrap_or_default().deaths;
        data
    }
}
