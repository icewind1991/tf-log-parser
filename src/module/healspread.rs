use crate::common::{SteamId3, SubjectId};
use crate::event::GameEvent;
use crate::module::EventHandler;
use crate::raw_event::RawEventType;
use crate::SubjectMap;
use std::collections::HashMap;
use std::convert::{Infallible, TryFrom};

#[derive(Default)]
pub struct HealSpreadHandler(HashMap<SteamId3, HashMap<SteamId3, u32>>);

impl EventHandler for HealSpreadHandler {
    type Output = HashMap<SteamId3, HashMap<SteamId3, u32>>;
    type Error = Infallible;

    fn does_handle(&self, ty: RawEventType) -> bool {
        matches!(ty, RawEventType::Healed)
    }

    fn handle(
        &mut self,
        _time: u32,
        subject: SubjectId,
        event: &GameEvent,
    ) -> Result<(), Self::Error> {
        let healer_steam_id = if let Some(steam_id) = subject.steam_id() {
            steam_id
        } else {
            return Ok(());
        };
        if let GameEvent::Healed(heal_event) = event {
            if let Ok(target_subject) = SubjectId::try_from(&heal_event.target) {
                if let Some(target_steam_id) = target_subject.steam_id() {
                    let healed = self
                        .0
                        .entry(SteamId3(healer_steam_id))
                        .or_default()
                        .entry(SteamId3(target_steam_id))
                        .or_default();
                    *healed += heal_event.amount
                }
            }
        }
        Ok(())
    }

    fn finish(self, _subjects: &SubjectMap) -> Self::Output {
        self.0
    }
}
