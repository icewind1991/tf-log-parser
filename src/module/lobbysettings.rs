use crate::common::SubjectId;
use crate::event::GameEvent;
use crate::module::EventHandler;
use crate::raw_event::RawEventType;
use crate::SubjectMap;
use chrono::{DateTime, FixedOffset, NaiveDateTime, TimeZone, Utc};
use std::str::{FromStr, ParseBoolError};
use steamid_ng::SteamID;
use thiserror::Error;

#[derive(Debug)]
pub enum GameType {
    Sixes,
    Highlander,
}

impl FromStr for GameType {
    type Err = LobbySettingsError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "6v6" => Ok(Self::Sixes),
            "highlander" => Ok(Self::Highlander),
            unknown => Err(LobbySettingsError::UnknownGameType(unknown.into())),
        }
    }
}

#[derive(Debug)]
pub enum Location {
    Europe,
    NorthAmerica,
}

impl FromStr for Location {
    type Err = LobbySettingsError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "Europe" => Ok(Self::Europe),
            "North America" => Ok(Self::NorthAmerica),
            unknown => Err(LobbySettingsError::UnknownLocation(unknown.into())),
        }
    }
}

#[derive(Debug, Default)]
pub struct LobbyLeader {
    name: String,
    steam_id: SteamID,
}

impl FromStr for LobbyLeader {
    type Err = LobbySettingsError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if let Some((name, steam_id)) = s.rsplit_once(" (") {
            if let Ok(steam_id) = steam_id.trim_end_matches(")").parse::<u64>() {
                Ok(LobbyLeader {
                    name: name.into(),
                    steam_id: steam_id.into(),
                })
            } else {
                Err(LobbySettingsError::MalformedLeader(s.into()))
            }
        } else {
            Err(LobbySettingsError::MalformedLeader(s.into()))
        }
    }
}

#[derive(Debug)]
pub struct Settings {
    id: u32,
    leader: LobbyLeader,
    map: String,
    game_type: GameType,
    location: Location,
    advanced: bool,
    region_lock: bool,
    allow_offclassing: bool,
    balancing: bool,
    restriction: String,
    mumble_required: bool,
    date: DateTime<Utc>,
    server: String,
}

impl Default for Settings {
    fn default() -> Self {
        Settings {
            id: 0,
            leader: LobbyLeader::default(),
            map: "".to_string(),
            game_type: GameType::Sixes,
            location: Location::Europe,
            advanced: false,
            region_lock: false,
            allow_offclassing: false,
            balancing: false,
            restriction: "".to_string(),
            mumble_required: false,
            date: Utc.ymd(1, 1, 1).and_hms(0, 0, 0),
            server: "".to_string(),
        }
    }
}

#[derive(Debug, Error)]
pub enum LobbySettingsError {
    #[error("Malformed lobby id: {0}")]
    InvalidLobbyId(String),
    #[error("Unknown game type: {0}")]
    UnknownGameType(String),
    #[error("Unknown location: {0}")]
    UnknownLocation(String),
    #[error("Unknown timezone in date: {0}")]
    UnknownTimezone(String),
    #[error("Malformed leader: {0}")]
    MalformedLeader(String),
    #[error("{0}")]
    InvalidBool(#[from] ParseBoolError),
    #[error("{0}")]
    InvalidDate(#[from] chrono::ParseError),
}

#[derive(Default)]
pub struct LobbySettingsHandler(Settings);

impl EventHandler for LobbySettingsHandler {
    type Output = Option<Settings>;
    type Error = LobbySettingsError;

    fn does_handle(&self, ty: RawEventType) -> bool {
        matches!(ty, RawEventType::Say)
    }

    fn handle(
        &mut self,
        _time: u32,
        subject: SubjectId,
        event: &GameEvent,
    ) -> Result<(), Self::Error> {
        if !matches!(subject, SubjectId::Console) {
            return Ok(());
        }
        if let GameEvent::Say(msg) = event {
            if let Some((id, _)) = msg
                .strip_prefix("TF2Center Lobby #")
                .and_then(|s| str::split_once(s, " |"))
            {
                self.0.id = id
                    .parse()
                    .map_err(|_| LobbySettingsError::InvalidLobbyId(id.into()))?;
            }
            if let Some((key, value)) = msg.split_once(": ") {
                match key {
                    "Leader" => self.0.leader = value.parse()?,
                    "Map" => self.0.map = value.into(),
                    "GameType" => self.0.game_type = value.parse()?,
                    "Location" => self.0.location = value.parse()?,
                    "Advanced Lobby" => self.0.advanced = value.parse()?,
                    "Region lock" => self.0.region_lock = value.parse()?,
                    "Allow offclassing" => self.0.allow_offclassing = value.parse()?,
                    "Balancing" => self.0.balancing = value.parse()?,
                    "Restriction" => self.0.restriction = value.into(),
                    "Mumble required" => self.0.mumble_required = value.parse()?,
                    "Launch date" => {
                        self.0.date = get_timezone(value)?
                            .from_local_datetime(&NaiveDateTime::parse_from_str(
                                value,
                                "%a %b %d %H:%M:%S %Z %Y",
                            )?)
                            .earliest()
                            .unwrap()
                            .into()
                    }
                    "Server" => self.0.server = value.into(),
                    _ => {}
                }
            }
        }
        Ok(())
    }

    fn finish(self, _subjects: &SubjectMap) -> Self::Output {
        if self.0.id > 0 {
            Some(self.0)
        } else {
            None
        }
    }
}

fn get_timezone(date: &str) -> Result<FixedOffset, LobbySettingsError> {
    if date.contains("CEST") {
        Ok(FixedOffset::east(2 * 60 * 60))
    } else if date.contains("CET") {
        Ok(FixedOffset::east(60 * 60))
    } else {
        Err(LobbySettingsError::UnknownTimezone(date.into()))
    }
}
