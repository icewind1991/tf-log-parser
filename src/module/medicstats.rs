use crate::common::SubjectId;
use crate::event::GameEvent;
use crate::module::EventHandler;
use crate::raw_event::RawEventType;
use crate::{SubjectData, SubjectMap};
use serde::Serialize;
use thiserror::Error;

#[derive(Default)]
pub struct MedicStatsBuilder {
    advantages_lost: u32,
    biggest_advantage_lost: f32,
    near_full_charge_death: u32,
    deaths_after_uber: u32,
    total_time_before_healing: f32,
    start_healing_count: u32,
    total_time_to_build: u32,
    uber_build_count: u32,
    total_time_to_use: f32,
    total_uber_length: f32,
    charge_count: u32,
    last_build_start: u32,
    last_uber_end: u32,
    drops: u32,
}

#[derive(Debug, Serialize)]
pub struct MedicStats {
    advantages_lost: u32,
    biggest_advantage_lost: f32,
    near_full_charge_death: u32,
    deaths_after_uber: u32,
    avg_time_before_healing: f32,
    avg_time_to_build: f32,
    avg_time_to_use: f32,
    avg_uber_length: f32,
    charge_count: u32,
    drops: u32,
}

impl From<MedicStatsBuilder> for MedicStats {
    fn from(builder: MedicStatsBuilder) -> Self {
        MedicStats {
            advantages_lost: builder.advantages_lost,
            biggest_advantage_lost: builder.biggest_advantage_lost,
            near_full_charge_death: builder.near_full_charge_death,
            deaths_after_uber: builder.deaths_after_uber,
            avg_time_before_healing: builder.total_time_before_healing
                / builder.start_healing_count as f32,
            avg_time_to_build: builder.total_time_to_build as f32 / builder.uber_build_count as f32,
            avg_time_to_use: builder.total_time_to_use / builder.charge_count as f32,
            avg_uber_length: builder.total_uber_length / builder.charge_count as f32,
            charge_count: builder.charge_count,
            drops: builder.drops,
        }
    }
}

#[derive(Error, Debug)]
#[error("Invalid charge event: {0}")]
pub struct InvalidMedicEvent(String);

#[derive(Default)]
pub struct MedicStatsHandler;

impl EventHandler for MedicStatsHandler {
    type GlobalOutput = ();
    type PerSubjectData = MedicStatsBuilder;
    type PerSubjectOutput = MedicStats;

    fn does_handle(&self, ty: RawEventType) -> bool {
        matches!(
            ty,
            RawEventType::ChargeDeployed
                | RawEventType::ChargeEnd
                | RawEventType::ChargeReady
                | RawEventType::UberAdvantageLost
                | RawEventType::MedicDeath
                | RawEventType::FirstHealAfterSpawn
        )
    }

    fn handle(
        &mut self,
        time: u32,
        _subject: SubjectId,
        subject_data: &mut Self::PerSubjectData,
        event: &GameEvent,
    ) {
        match event {
            GameEvent::ChargeEnded(end) => {
                subject_data.total_uber_length += end.duration.unwrap_or_default();
                subject_data.last_uber_end = time;
            }
            GameEvent::ChargeDeployed(_deployed) => {
                subject_data.charge_count += 1;
            }
            GameEvent::AdvantageLost(lost) => {
                subject_data.advantages_lost += 1;
                let time = lost.time.unwrap_or_default();
                if time > subject_data.biggest_advantage_lost {
                    subject_data.biggest_advantage_lost = time;
                }
            }
            GameEvent::FirstHeal(first) => {
                subject_data.total_time_before_healing += first.time.unwrap_or_default();
                subject_data.start_healing_count += 1;
                subject_data.last_build_start = time;
            }
            GameEvent::ChargeReady => {
                if subject_data.last_build_start > 0 {
                    let build_time = time - subject_data.last_build_start;
                    subject_data.last_build_start = 0;
                    subject_data.total_time_to_build += build_time;
                    subject_data.uber_build_count += 1;
                }
            }
            GameEvent::MedicDeath(death) => {
                let charge = death.charge.unwrap_or_default();
                if charge >= 95 && charge < 100 {
                    subject_data.near_full_charge_death += 1;
                } else if charge >= 100 {
                    subject_data.drops += 1;
                }
                if time - subject_data.last_uber_end <= 10 {
                    subject_data.deaths_after_uber += 1;
                }
            }
            _ => {}
        }
    }

    fn finish_global(self, _subjects: &SubjectMap) -> Self::GlobalOutput {
        ()
    }

    fn finish_per_subject(
        &self,
        _subject: &SubjectData,
        data: Self::PerSubjectData,
    ) -> Self::PerSubjectOutput {
        data.into()
    }
}
