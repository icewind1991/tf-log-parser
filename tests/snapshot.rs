use serde::Serialize;
use std::collections::BTreeMap;
use std::fs::read_to_string;
use test_case::test_case;
use tf_log_parser::module::{ClassStats, MedicStats};
use tf_log_parser::{parse, EventHandler, LogHandler, LogHandlerPerSubjectOutput};

#[derive(Serialize)]
struct LogResult {
    global: <LogHandler as EventHandler>::GlobalOutput,
    per_player: BTreeMap<String, LogPlayerData>,
}

#[derive(Serialize)]
struct LogPlayerData {
    stats: ClassStatsRaw,
    heals: BTreeMap<String, u32>,
    medic: MedicStats,
}

impl From<LogHandlerPerSubjectOutput> for LogPlayerData {
    fn from(value: LogHandlerPerSubjectOutput) -> Self {
        LogPlayerData {
            stats: value.class_stats.into(),
            medic: value.medic_stats,
            heals: value
                .heal_spread
                .into_iter()
                .map(|(user, heals)| (user.0.steam3(), heals))
                .collect(),
        }
    }
}

#[derive(Serialize)]
struct ClassStatsRaw {
    kills: [u16; 10],
    deaths: [u16; 10],
    assists: [u16; 10],
    damage: [u16; 10],
}

impl From<ClassStats> for ClassStatsRaw {
    fn from(value: ClassStats) -> Self {
        ClassStatsRaw {
            kills: value.kills.into(),
            deaths: value.deaths.into(),
            assists: value.assists.into(),
            damage: value.damage.into(),
        }
    }
}

#[test_case("log_6s.log")]
#[test_case("log_2788889.log")]
#[test_case("log_2892242.log")]
#[test_case("log_bball.log")]
#[test_case("log_hl.log")]
fn test_parse(name: &str) {
    let content = read_to_string(&format!("test_data/{}", name)).unwrap();
    let (global, per_player) = parse(&content).unwrap();
    let log = LogResult {
        global,
        per_player: per_player
            .into_iter()
            .map(|(key, value)| (key.0.steam3(), value.into()))
            .collect(),
    };
    insta::assert_json_snapshot!(format!("parse-{}", name), log);
}
