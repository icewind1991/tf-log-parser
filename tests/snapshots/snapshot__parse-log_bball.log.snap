---
source: tests/snapshot.rs
expression: log
---
{
  "global": {
    "chat": [
      {
        "time": 0,
        "name": "Bv",
        "steam_id": 76561198048943710,
        "message": "[P-REC] Stop record.",
        "chat_type": "Team"
      },
      {
        "time": 0,
        "name": "Charon",
        "steam_id": 76561198061714753,
        "message": "[P-REC] Stop record.",
        "chat_type": "Team"
      },
      {
        "time": 53,
        "name": "Bv",
        "steam_id": 76561198048943710,
        "message": "[P-REC] Recording...",
        "chat_type": "Team"
      },
      {
        "time": 53,
        "name": "Charon",
        "steam_id": 76561198061714753,
        "message": "[P-REC] Recording...",
        "chat_type": "Team"
      }
    ]
  },
  "per_player": {
    "[U:1:101449025]": {
      "stats": {
        "kills": [
          0,
          25,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "deaths": [
          0,
          32,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          0,
          9,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "damage": [
          0,
          5039,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:101608870]": {
      "stats": {
        "kills": [
          0,
          31,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "deaths": [
          0,
          28,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          0,
          10,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "damage": [
          0,
          8262,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:82123920]": {
      "stats": {
        "kills": [
          0,
          31,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "deaths": [
          0,
          29,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          0,
          6,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "damage": [
          0,
          6461,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    },
    "[U:1:88677982]": {
      "stats": {
        "kills": [
          0,
          30,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "deaths": [
          0,
          28,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "assists": [
          0,
          8,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ],
        "damage": [
          0,
          5764,
          0,
          0,
          0,
          0,
          0,
          0,
          0,
          0
        ]
      },
      "heals": {},
      "medic": {
        "advantages_lost": 0,
        "biggest_advantage_lost": 0.0,
        "near_full_charge_death": 0,
        "deaths_after_uber": 0,
        "avg_time_before_healing": 0.0,
        "avg_time_to_build": 0.0,
        "avg_time_to_use": 0.0,
        "avg_uber_length": 0.0,
        "charge_count": 0,
        "drops": 0
      }
    }
  }
}
